wgpu-hal/src/dx12/device.rs

```diff
@@ -221,62 +222,151 @@ impl super::Device {
+        // hassle_rs doesn't check for `dxil.dll` if you only call `hassle_rs::compile_hlsl` but calling hassle_rs::Dxil::new will .
+        profiling::scope!("hassle_rs::compile_hlsl");
+        let (result, log_level) = match hassle_rs::Dxil::new(None).and_then(|_| {
+            hassle_rs::compile_hlsl(
                 source_name,
-                ptr::null(),
-                ptr::null_mut(),
-                raw_ep.as_ptr(),
-                full_stage.as_ptr() as *const i8,
-                compile_flags,
-                0,
-                shader_data.mut_void() as *mut *mut _,
-                error.mut_void() as *mut *mut _,
+                &source,
+                /*this is safe because `raw_ep` was converted from `String`*/
+                raw_ep.to_str().unwrap(),
+                &full_stage,
+                &compile_flags,
+                &[],
             )
-        };
+        }) {
+            Ok(dxil) => (Ok(super::ShaderDXIL::Dxc(dxil)), log::Level::Info),
+            Err(err) => match err {
+                hassle_rs::HassleError::Win32Error(err) => {
+                    let err_str = i32::into_result(err.0).unwrap_err();
+                    (
+                        Err(crate::PipelineError::Linkage(
+                            stage_bit,
+                            format!(
+                                "hassle_rs::compile_hlsl failed with error:({}) :{}",
+                                err, err_str
+                            ),
+                        )),
+                        log::Level::Warn,
+                    )
+                }
+                hassle_rs::HassleError::WindowsOnly(err) => (
+                    Err(crate::PipelineError::Linkage(
+                        stage_bit,
+                        format!("hassle_rs::compile_hlsl failed with error: {}", err),
+                    )),
+                    log::Level::Warn,
+                ),
 
-        let (result, log_level) = match hr.into_result() {
-            Ok(()) => (Ok(shader_data), log::Level::Info),
-            Err(e) => {
-                let mut full_msg = format!("D3DCompile error ({})", e);
-                if !error.is_null() {
-                    use std::fmt::Write as _;
-                    let message = unsafe {
-                        slice::from_raw_parts(
-                            error.GetBufferPointer() as *const u8,
-                            error.GetBufferSize(),
-                        )
-                    };
-                    let _ = write!(full_msg, ": {}", String::from_utf8_lossy(message));
-                    unsafe {
-                        error.destroy();
-                    }
+                hassle_rs::HassleError::LoadLibraryError { filename, inner: _ } => {
+                    log::warn!("{:?} was not found ,falling back to FXC ,in order to use the new hlsl compiler `DXC`,`dxcompiler.dll` and `dxil.dll` should be included in the current execution environment" ,filename );
```

Comment:
We should only send this warning once per program execution.

Reply:
And what's up with the commas?